File: Visualizer/src/render/grid-renderer.js
import { Style } from '../core/types.js';

/**
 * Renders the orthogonal grid for two directed segments.
 * - Horizontal colored lines in the V-segment's real region
 * - Vertical colored lines in the H-segment's real region
 * - Yellow fill in the quadrant where both are imaginary
 */
export class GridRenderer {
  /**
   * @param {import('../render/svg-canvas.js').SvgCanvas} canvas
   */
  constructor(canvas) {
    this.canvas = canvas;
    this.group = canvas.createGroup();
  }

  /**
   * Update the grid based on current segment values.
   * @param {object} hSeg - Horizontal segment { imaginary, real }
   * @param {object} vSeg - Vertical segment { imaginary, real }
   * @param {object} hColors - Horizontal segment colors { grid }
   * @param {object} vColors - Vertical segment colors { grid }
   */
  update(hSeg, vSeg, hColors, vColors) {
    const { canvas } = this;
    const g = this.group;

    // Clear
    while (g.firstChild) g.removeChild(g.firstChild);

    const sw = Style.gridStrokeWidth;
    const spacing = Style.gridSpacing;

    // Full extents in math space
    const xMin = hSeg.imaginary;
    const xMax = hSeg.real;
    const yMin = vSeg.imaginary;
    const yMax = vSeg.real;

    // Pixel bounds
    const topLeft = canvas.mathToPixel(xMin, yMax);
    const botRight = canvas.mathToPixel(xMax, yMin);

    // --- Yellow fill: quadrant where both are imaginary (x < 0 AND y < 0) ---
    const origin = canvas.mathToPixel(0, 0);
    const imagCorner = canvas.mathToPixel(xMin, yMin);

    if (xMin < 0 && yMin < 0) {
      const fillX = imagCorner.x;
      const fillY = origin.y; // y=0 in pixels (top of the imaginary-Y region since Y is flipped from the math-positive above)
      const fillW = origin.x - imagCorner.x;
      const fillH = imagCorner.y - origin.y;
      if (fillW > 0 && fillH > 0) {
        g.appendChild(canvas.createRect(fillX, fillY, fillW, fillH, {
          fill: Style.fillColor, stroke: 'none'
        }));
      }
    }

    // --- Horizontal grid lines (V-segment's color, in V's real region) ---
    // These are horizontal lines spanning the full X range of H-segment
    // Drawn for each integer Y in V's real region (y > 0 up to yMax)
    {
      const startY = Math.max(Math.ceil(0), Math.ceil(yMin));
      const endY = Math.floor(yMax);
      for (let my = startY; my <= endY; my++) {
        if (my === 0) continue; // skip origin line
        const left = canvas.mathToPixel(xMin, my);
        const right = canvas.mathToPixel(xMax, my);
        g.appendChild(canvas.createLine(
          left.x, left.y, right.x, right.y,
          { stroke: vColors.grid, 'stroke-width': sw }
        ));
      }
    }

    // --- Vertical grid lines (H-segment's color, in H's real region) ---
    // These are vertical lines spanning the full Y range of V-segment
    // Drawn for each integer X in H's real region (x > 0 up to xMax)
    {
      const startX = Math.max(Math.ceil(0), Math.ceil(xMin));
      const endX = Math.floor(xMax);
      for (let mx = startX; mx <= endX; mx++) {
        if (mx === 0) continue; // skip origin line
        const top = canvas.mathToPixel(mx, yMax);
        const bot = canvas.mathToPixel(mx, yMin);
        g.appendChild(canvas.createLine(
          top.x, top.y, bot.x, bot.y,
          { stroke: hColors.grid, 'stroke-width': sw }
        ));
      }
    }
  }
}

File: Visualizer.WinForms/Rendering/GridRenderer.cs
using OpenTK.Graphics;
using ResoEngine.Visualizer.Core;
using SkiaSharp;
using static System.Windows.Forms.AxHost;

namespace ResoEngine.Visualizer.Rendering;

/// <summary>
/// Renders the orthogonal grid for two directed segments:
/// - Horizontal lines in V-segment's real region (blue tint)
/// - Vertical lines in H-segment's real region (red tint)
/// - Yellow fill in the double-negative (imaginary) quadrant
/// </summary>
public class GridRenderer : IDisposable
{
    private readonly CoordinateSystem _coords;
    private readonly SKPaint _fillPaint;

    public GridRenderer(CoordinateSystem coords)
    {
        _coords = coords;

        _fillPaint = new SKPaint
        {
            Style = SKPaintStyle.Fill,
            Color = VisualStyle.FillColor,
            IsAntialias = false,
        };
    }

    float _xMin;
    float _xMax;
    float _yMin;
    float _yMax;
    public void Render(SKCanvas canvas, DirectedSegment hSeg, DirectedSegment vSeg,
                       SegmentColorSet hColors, SegmentColorSet vColors)
    {
        _xMin = hSeg.Imaginary;
        _xMax = hSeg.Real;
        _yMin = vSeg.Imaginary;
        _yMax = vSeg.Real;

        var origin = _coords.MathToPixel(0, 0);
        var imagCorner = _coords.MathToPixel(_xMin, _yMin);

        float fillX = imagCorner.X;
        float fillY = origin.Y;
        float fillW = origin.X - imagCorner.X;
        float fillH = imagCorner.Y - origin.Y;

        canvas.DrawRect(fillX, fillY, fillW, fillH, _fillPaint);

        // --- Horizontal grid lines (V-segment's color, in V's real region) ---
        using (var bluePaint = GetPen(vColors.Grid))
        {
            var minVal = MathF.Min(_yMin, _yMax);
            var maxVal = MathF.Max(_yMin, _yMax);
            int start = (int)Math.Ceiling(minVal);
            int end = (int)Math.Floor(maxVal);
            for (int my = start; my <= end; my++)
            {
                DrawVLine(my, hSeg, bluePaint, canvas);
            }
        }

        // --- Vertical grid lines (H-segment's color, in H's real region) ---
        using (var redPaint = GetPen(hColors.Grid))
        {
            var minVal = MathF.Min(_xMin, _xMax);
            var maxVal = MathF.Max(_xMin, _xMax);
            int start = (int)Math.Ceiling(minVal);
            int end = (int)Math.Floor(maxVal);
            for (int mx = start; mx <= end; mx++)
            {
                DrawHLine(mx, vSeg, redPaint, canvas);
            }
        }
    }


    private void DrawVLine(float my, DirectedSegment seg, SKPaint bluePaint, SKCanvas canvas)
    {
        if (my == 0) return;
        var hMin = my > 0 ? MathF.Min(_xMin, 0) : MathF.Max(_xMin, 0);
        var start = _coords.MathToPixel(hMin, my);
        var end = _coords.MathToPixel(_xMax, my);
        canvas.DrawLine(start, end, bluePaint);
    }
    private void DrawHLine(float mx, DirectedSegment seg, SKPaint redPaint, SKCanvas canvas)
    {
        if (mx == 0) return;
        var vMin = mx > 0 ? MathF.Min(_yMin, 0) : MathF.Max(_yMin, 0);
        var start = _coords.MathToPixel(mx, vMin);
        var end = _coords.MathToPixel(mx, _yMax);
        canvas.DrawLine(start, end, redPaint);
    }
    private SKPaint GetPen(SKColor color)
    {
        return new SKPaint
        {
            Style = SKPaintStyle.Stroke,
            StrokeWidth = VisualStyle.GridStrokeWidth,
            Color = color,
            IsAntialias = true,
        };
    }
    public void Dispose()
    {
        _fillPaint.Dispose();
        GC.SuppressFinalize(this);
    }
}
